server/rating_service/rating_service.py

```diff
@@ -148,192 +143,353 @@ async def _handle_rating_queue(self) -> None:
         self._logger.debug("RatingService stopped.")
 
     async def _rate(self, summary: GameRatingSummary) -> None:
-        rating_data = await self._get_rating_data(summary)
-        new_ratings = GameRater.compute_rating(rating_data)
-
-        outcome_map = {
-            player_id: team.outcome
-            for team in summary.teams
-            for player_id in team.player_ids
-        }
+        async with self._db.acquire() as conn:
+            player_ids = [
+                player_id
+                for team in summary.teams
+                for player_id in team.player_ids
+            ]
+            # Fetch all players rating info from the database
+            player_ratings = await self._get_all_player_ratings(
+                conn, player_ids
+            )
+            # Initialize the ratings we need
+            old_ratings = await self._get_players_rating_initialized(
+                conn, player_ratings, summary.rating_type
+            )
+            # Create rating data object
+            rating_data = self._get_rating_data(conn, summary, old_ratings)
+            new_ratings = GameRater.compute_rating(rating_data)
+
+            outcome_map = {
+                player_id: team.outcome
+                for team in summary.teams
+                for player_id in team.player_ids
+            }
 
-        old_ratings = {
-            player_id: rating
-            for team in rating_data
-            for player_id, rating in team.ratings.items()
-        }
-        await self._persist_rating_changes(
-            summary.game_id,
+            await self._persist_rating_changes(
+                conn,
+                summary.game_id,
+                summary.rating_type,
+                old_ratings,
+                new_ratings,
+                outcome_map
+            )
+        await self._publish_rating_changes(
             summary.rating_type,
             old_ratings,
             new_ratings,
             outcome_map
         )
+
+        # TODO: If we add hidden ratings, make sure to check for them here.
+        # Hidden ratings should not affect global.
+        if summary.rating_type == RatingType.GLOBAL:
+            return
+
+        # Now perform global rating adjustments.
+        async with self._db.acquire() as conn:
```

Comment:
I think I'm going to leave it as is so that we're not holding the database connection open while posting to rabbitmq. Not sure if it really matters, but it just seems like a good idea to me to drop the connection while we're not using it and instead waiting on rabbitmq messages and then acquire a new connection after when we do the global rating adjustment.